server/world/mcdb/provider.go

```diff
@@ -141,6 +144,76 @@ func (p *Provider) SaveSettings(s *world.Settings) {
 	p.saveDifficulty(s.Difficulty)
 }
 
+// LoadPlayerSpawnPosition loads the players spawn position stored in the level.dat from their UUID.
+func (p *Provider) LoadPlayerSpawnPosition(uuid uuid.UUID) (pos mgl64.Vec3, err error) {
+	data, err := p.db.Get(append([]byte("player_"), uuid[:]...), nil)
+	if err != nil {
+		return mgl64.Vec3{}, err
+	}
+	var playerData map[string]string
+
+	if err := nbt.UnmarshalEncoding(data, &playerData, nbt.LittleEndian); err != nil {
+		return mgl64.Vec3{}, err
+	}
+	if playerData["MsaID"] != uuid.String() && playerData["ServerId"] == "" {
+		return mgl64.Vec3{}, fmt.Errorf("LoadPlayerSpawn: Invalid data for player %s", uuid.String())
+	}
+	serverDb, err := p.db.Get([]byte(playerData["ServerId"]), nil)
+	if err != nil {
+		return mgl64.Vec3{}, err
+	}
+
+	var serverData map[string]any
+	if err := nbt.UnmarshalEncoding(serverDb, &serverData, nbt.LittleEndian); err != nil {
+		return mgl64.Vec3{}, err
+	}
+
+	x, y, z := serverData["SpawnX"], serverData["SpawnY"], serverData["SpawnZ"]
+	if x == nil || y == nil || z == nil {
+		return mgl64.Vec3{}, fmt.Errorf("LoadPlayerSpawn: player spawn position is non-existent")
+	}
+	return mgl64.Vec3{x.(float64), y.(float64), z.(float64)}, nil
+}
+
+// SavePlayerSpawnPosition saves the player spawn position passed to the level.dat.
+func (p Provider) SavePlayerSpawnPosition(uuid uuid.UUID, pos mgl64.Vec3) error {
+	data, err := p.db.Get(append([]byte("player_"), uuid[:]...), nil)
+	if errors.Is(err, leveldb.ErrNotFound) {
+		data, err = nbt.MarshalEncoding(map[string]string{
+			"MsaID":    uuid.String(),
+			"ServerID": "player_server_" + uuid.String(),
+		}, nbt.LittleEndian)
+		if err != nil {
+			return err
+		}
+		if err := p.db.Put(append([]byte("player_"), uuid[:]...), data, nil); err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+	var playerData map[string]string
+
+	if err := nbt.UnmarshalEncoding(data, &playerData, nbt.LittleEndian); err != nil {
+		return err
+	}
+	if playerData["MsaID"] != uuid.String() && playerData["ServerId"] == "" {
+		return fmt.Errorf("SavePlayerSpawn: Invalid data for player %s", uuid.String())
+	}
+	coords, err := nbt.MarshalEncoding(map[string]any{
```

Comment:
So there's one last thing I think you might consider changing here. Right now, we overwrite all data stored with just the spawn X, spawn Y and spawn Z. Maybe it's an idea to first load the map of data (as done in `LoadPlayerSpawnPosition`), then overwrite only the `SpawnX`, `SpawnY` and `SpawnZ` fields and write the map again. That way we don't lose all the other fields that are still stored there, which means better compatibility with vanilla.